### PR TITLE
add restartCountLimit for pod

### LINKED ISSUE #49466: Kubernetes how to limit the pods restart count
**In kubernetes, how can I limit the pods restart count ?** 

Now, when the pods exit with none zero or other reasons ,then it will be restarted according to restartPolicy (Always、OnFailure、Never).

From the source code, there is a property (restart) to add the restart count. I think that when the property equals to the number I expect to restart count, then the pod will be deleted. Weather it will work? Or there are other solutions to work this problem? Thank you !!!

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
When the pods exit with none zero or other reasons ,then it will be restarted according to restartPolicy. For pods with `Always` restart policy, there should a way to stop such infinite restart loop, which is wasting the resources and may cause undesired restarts.

This PR adds a new annotation `"io.kubernetes.pod.restartCountLimit"` for pod to limit pod restart count.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #49466

**Special notes for your reviewer**:
/sig node
/kind featue

**Release note**:

```release-note
add an annotation to limit pods restart count
```

### CODE REVIEW

**Comment by feiskyer** [comment]
> may cause undesired restarts.

Sorry, I didn't get this. What do you mean by this?

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -732,6 +736,28 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStat
...
+				} else {
+					currentRestartCount := 0
+					for _, containerStatus := range podStatus.ContainerStatuses {
+						currentRestartCount += containerStatus.RestartCount
+					}
+					if currentRestartCount > restartCountLimit {
+						msg := fmt.Sprintf("Stop restarting container %+v in pod %v due to exceeding pod restart count limit", container, format.Pod(pod))
+						startContainerResult.Fail(errors.New("ExceededRestartCountLimit"), msg)
+						glog.V(4).Infof(msg)
+						continue
```
Comment (tianshapjq) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: doesn't it need to break and stop looping the rest containers?

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 2]
```diff
@@ -732,6 +736,28 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStat
...
+				restartCountLimit, err := strconv.Atoi(restartCountLimitStr)
+				if err != nil {
+					glog.V(5).Infof("Failed to parse restart count limit in pod %v: %+v", format.Pod(pod), err)
+				} else {
+					currentRestartCount := 0
+					for _, containerStatus := range podStatus.ContainerStatuses {
+						currentRestartCount += containerStatus.RestartCount
+					}
+					if currentRestartCount > restartCountLimit {
+						msg := fmt.Sprintf("Stop restarting container %+v in pod %v due to exceeding pod restart count limit", container, format.Pod(pod))
```
Comment (tianshapjq) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: %v -> %s

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 3]
```diff
@@ -732,6 +736,28 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStat
 			continue
 		}
 
+		// limit pod restart count if set
+		restartCountLimitStr, ok := pod.Annotations[podRestartCountLimitAnn]
+		if ok {
+			if len(restartCountLimitStr) > 0 {
+				restartCountLimit, err := strconv.Atoi(restartCountLimitStr)
+				if err != nil {
+					glog.V(5).Infof("Failed to parse restart count limit in pod %v: %+v", format.Pod(pod), err)
```
Comment (tianshapjq) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: would it be better to put it into the event pool of pod? then I can describe pod to see what happens to this pod.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 4]
```diff
@@ -732,6 +736,28 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStat
...
+				} else {
+					currentRestartCount := 0
+					for _, containerStatus := range podStatus.ContainerStatuses {
+						currentRestartCount += containerStatus.RestartCount
+					}
+					if currentRestartCount > restartCountLimit {
+						msg := fmt.Sprintf("Stop restarting container %+v in pod %v due to exceeding pod restart count limit", container, format.Pod(pod))
+						startContainerResult.Fail(errors.New("ExceededRestartCountLimit"), msg)
```
Comment (mlmhl) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: What's the status of Pod when `restartCountLimit` exceeded? Do we need to make it as `Failed`?

**Review by yujuhong — CHANGES_REQUESTED** [review]
@dixudx I think this may need a (short) proposal detailing what the desired behavior is, why is this set per pod (vs per node) and what goals the change is trying to achieve.

Besides, annotations are discouraged and API chaneg should be considered with feature gates if this is necessary.

**Comment by michaelfavia** [comment]
As someone affected by this issue and found this PR looking for a builtin solution I can chime in here as well.

We run pods as part of a HPA and occasionally some of the pods get stuck in this restart loop. The cause varies from what I can tell. Maybe its due to initial filesystem instability during a initContainer step that copies the resources to the disk or some other reason. 

When this happens a simple manual deletion of the pod causes it to be instantly restored by the HPA/RS and is back to fully functioning. 

Perhaps this annotation is best put on the HPA (or ReplicaSet - the pod manager) because I struggle to think of a use case for restartCountLimit on a Pod of which I think the whole point is its resilience. It probably should be responsible for its own death in the same vein of it not being responsible for its creation. ;)

@yujuhong: Does this explanation help at all?

**Comment by yujuhong** [comment]
@michaelfavia thanks for the explanation. I could think of ways that the feature would be useful, but I have some concerns about how it's implemented in the PR.

As you said, HPA or other controllers could detect the crashlooping, delete & re-create the pods based on it's only criteria. It might be an issue figuring out which controller should own this logic, and that's where implementing this in kubelet *might* be preferable. Another use case would be saving resources on the node, although with the current backoff mechanism, I haven't seen this being a performance bottleneck yet. Overall, I  think this deserves more discussion, and adding new annotations for kubelet to check and implement is probably not the way to go :-)

**Comment by mlmhl** [comment]
Agree with @yujuhong , we need more discussion of this feature, such as:

1. Some use case or user story, this can help us better define our demand.
2. What should we do when `restartCountLimit` exceeded? Just skip creating the container or make the `Pod` failed, so that according controllers can re-create it.

**Pull request closed without merging**